_posts/defining-messages.md

```diff
@@ -0,0 +1,169 @@
+
+## POCO
+
+One could say (and I'm one of them), "Keep it simple, stupid!" ([KISS](https://en.wikipedia.org/wiki/KISS_principle)) and use POCO classes. For example:
+
+```csharp
+public class OrderAccepted
+{
+   public string OrderNumber { get; set; }
+}
+```
+
+The presented class represents an event in a system dealing with orders. 
+The presented class comes with a few potential flaws that are worth listing (in no particular order):
+
+1. The `OrderAccepted` type is mutable. After deserializing the message, a receiver can change the `OrderNumber` property value. And that might even be worse because an event must be immutable (for the purists), considering it represents the past.
+2. [Contracts evolution](https://milestone.topics.it/2022/07/04/messages-evolution.html) might be tricky because using classes makes adopting a multiple inheritance approach difficult if not impossible.
+
+A quick solution to the first problem could be to use a private setter for the public properties, or similar approaches, and a constructor that allows setting those property values once at initialization time:
+
+```csharp
+public class OrderAccepted
+{
+   public OrderAccepted( string orderNumber ) => OrderNumber = orderNumber;
+   public string OrderNumber { get; private set; }
+}
+```
+
+That's neat, even if it's more verbose and might limit the serializer choice: We can only select serializers supporting setting private or read-only properties and capable of using non-default or private constructors.
+
+## Interfaces
+
+What sounds like a better way to define the `OrderAccepted` event is by using an interface.
+
+Messaging libraries like [NServiceBus support sending messages as interfaces](https://docs.particular.net/nservicebus/messaging/messages-as-interfaces) primarily to support multiple inheritance. At runtime, the library creates a dynamic proxy class that implements the interface, allowing to have code like the following:
+
+```csharp
+await endpoint.Send<IMyMessage>(message =>
+{
+    message.SomeProperty = "Hello world";
+}).ConfigureAwait(false);
+```
+
+In the above snippet, the `message` instance is the proxy type instance presented to the code as `IMyMessage`.
+
+With that in mind, can we define the `OrderAccepted` event as follows?
+
+```csharp
+public interface OrderAccepted
+{
+   string OrderNumber { get; }
+}
+```
+
+If we could, we'd have solved both the mutability issue of the POCO class and we'd be supporting multiple inheritance. Unfortunately, we cannot. If we were to use such an interface, the following code wouldn't compile:
+
+```csharp
+await endpoint.Publish<OrderAccepted>(message =>
+{
+    message.OrderNumber = "123ABC";
+}).ConfigureAwait(false);
+```
+
+The `OrderNumber` property is "get" only. It cannot be set. The interface works well for the message receiver rather than for the sender. The solution is to void creating the dynamic proxy class at the sender and instead have an explicit implementation:
```

Comment:
```suggestion
The `OrderNumber` property is "get" only. It cannot be set. The interface works well for the message receiver rather than for the sender. The solution is to avoid creating the dynamic proxy class at the sender and instead have an explicit implementation:
```